File: packages/server/src/routes/repo.py
from typing import List
from fastapi import APIRouter, Depends, Request, UploadFile, File, Form
from sqlalchemy.ext.asyncio import AsyncSession

from src.db.database import get_db
from src.db.models.user import User
from src.middlewares.auth import auth_middleware, optional_auth_middleware
from src.controllers import repo as repo_controller

router = APIRouter(prefix="/api/v1/repo", tags=["repo"])

@router.get("/{owner}/{name}/pull")
async def pull_from_remote(
    owner: str,
    name: str,
    local_head: str | None = None,
    current_user: User | None = Depends(optional_auth_middleware),
    db: AsyncSession = Depends(get_db),
):
    return await repo_controller.pull(
        owner, name, local_head, current_user, db
    )

@router.post("")
async def create_repository(
    name: str = Form(...),
    description: str = Form(None),
    visibility: str = Form("public"),
    current_user: User = Depends(auth_middleware),
    db: AsyncSession = Depends(get_db),
):
    return await repo_controller.create_repository(
        name, description, visibility, current_user, db
    )


@router.get("/user/{username}")
async def list_user_repositories(
    username: str,
    db: AsyncSession = Depends(get_db),
):
    return await repo_controller.list_user_repositories(username, db)


@router.get("/user/{username}/activity")
async def get_user_activity(
    username: str,
    db: AsyncSession = Depends(get_db),
    limit: int = 10,
):
    return await repo_controller.get_user_activity(username, db, limit)


@router.get("/user/{username}/stats")
async def get_user_stats(
    username: str,
    db: AsyncSession = Depends(get_db),
):
    return await repo_controller.get_user_stats(username, db)


@router.get("/user/{username}/contributions")
async def get_user_contributions(
    username: str,
    db: AsyncSession = Depends(get_db),
):
    return await repo_controller.get_user_contributions(username, db)


@router.get("/explore")
async def list_public_repositories(
    search: str | None = None,
    language: str | None = None,
    db: AsyncSession = Depends(get_db),
):
    return await repo_controller.list_public_repositories(db, search, language)


@router.get("/{owner}/{name}")
async def get_repository(
    owner: str,
    name: str,
    db: AsyncSession = Depends(get_db),
):
    return await repo_controller.get_repository(owner, name, db)


@router.get("/{owner}/{name}/tree/{branch}")
@router.get("/{owner}/{name}/tree/{branch}/{path:path}")
async def get_repository_tree(
    owner: str,
    name: str,
    branch: str,
    path: str = "",
    db: AsyncSession = Depends(get_db),
):
    return await repo_controller.get_repository_tree(owner, name, branch, path, db)


@router.get("/{owner}/{name}/blob/{branch}/{path:path}")
async def get_blob_content(
    owner: str,
    name: str,
    branch: str,
    path: str,
    db: AsyncSession = Depends(get_db),
):
    return await repo_controller.get_blob_content(owner, name, branch, path, db)


@router.get("/{owner}/{name}/commits/{branch}")
async def list_commits(
    owner: str,
    name: str,
    branch: str,
    page: int = 1,
    limit: int = 20,
    db: AsyncSession = Depends(get_db),
):
    return await repo_controller.list_commits(owner, name, branch, db, page, limit)


@router.get("/{owner}/{name}/commit/{hash}")
async def get_commit_detail(
    owner: str,
    name: str,
    hash: str,
    db: AsyncSession = Depends(get_db),
):
    return await repo_controller.get_commit_detail(owner, name, hash, db)


@router.get("/{owner}/{name}/commit/{hash}/diff")
async def get_commit_diff(
    owner: str,
    name: str,
    hash: str,
    db: AsyncSession = Depends(get_db),
):
    return await repo_controller.get_commit_diff(owner, name, hash, db)


@router.delete("/{owner}/{name}")
async def delete_repository(
    owner: str,
    name: str,
    current_user: User = Depends(auth_middleware),
    db: AsyncSession = Depends(get_db),
):
    return await repo_controller.delete_repository(owner, name, current_user, db)


@router.patch("/{owner}/{name}")
async def update_repository(
    owner: str,
    name: str,
    payload: dict,
    current_user: User = Depends(auth_middleware),
    db: AsyncSession = Depends(get_db),
):
    return await repo_controller.update_repository(
        owner,
        name,
        payload.get("name"),
        payload.get("description"),
        payload.get("visibility"),
        payload.get("default_branch"),
        current_user,
        db,
    )


@router.post("/{owner}/{name}/confirm-delete")
async def confirm_delete_repository(
    owner: str,
    name: str,
    payload: dict,
    current_user: User = Depends(auth_middleware),
    db: AsyncSession = Depends(get_db),
):
    return await repo_controller.confirm_delete_repository(
        owner, name, payload.get("confirmation_name"), current_user, db
    )


@router.get("/{owner}/{name}/branch/{branch}/head")
async def get_head_commit(
    owner: str,
    name: str,
    branch: str,
    db: AsyncSession = Depends(get_db),
):
    return await repo_controller.get_head_commit(owner, name, branch, db)


@router.post("/{owner}/{name}/push")
async def push(
    owner: str,
    name: str,
    metadata: str = Form(...),
    files: List[UploadFile] = File(default=[]),
    current_user: User = Depends(auth_middleware),
    db: AsyncSession = Depends(get_db),
):
    return await repo_controller.push(
        owner, name, metadata, files, current_user, db
    )


@router.get("/{owner}/{name}/pull")
async def pull(
    owner: str,
    name: str,
    local_head: str | None = None,
    current_user: User | None = Depends(optional_auth_middleware),
    db: AsyncSession = Depends(get_db),
):
    return await repo_controller.pull(owner, name, local_head, current_user, db)

File: frontend/src/api/repo.ts
import type { AxiosInstance } from 'axios'
import type { 
  Repository, 
  ApiResponse, 
  ActivityItem, 
  UserStats, 
  TreeEntry, 
  BlobResponse,
  Commit,
  DiffFile,
  PullRequest,
  PullRequestDetail
} from '@/types'

export interface BranchInfo {
  name: string
  headCommitHash: string | null
  updatedAt: string
}

export async function createRepository(
  client: AxiosInstance, 
  formData: FormData
) {
  const { data } = await client.post<ApiResponse<{ id: string; name: string }>>(
    '/api/v1/repo',
    formData,
    { headers: { 'Content-Type': 'multipart/form-data' } }
  )
  return data
}

export async function fetchUserRepositories(client: AxiosInstance, username: string) {
  const { data } = await client.get<ApiResponse<Repository[]>>(`/api/v1/repo/user-repos/${encodeURIComponent(username)}`)
  return data
}

export async function fetchUserActivity(client: AxiosInstance, username: string, limit = 10) {
  const { data } = await client.get<ApiResponse<ActivityItem[]>>(
    `/api/v1/repo/user-repos/${encodeURIComponent(username)}/activity?limit=${limit}`
  )
  return data
}

export async function fetchUserStats(client: AxiosInstance, username: string) {
  const { data } = await client.get<ApiResponse<UserStats>>(`/api/v1/repo/user-repos/${encodeURIComponent(username)}/stats`)
  return data
}

export interface ContributionData {
  contributions: Record<string, number>
  startDate: string
  endDate: string
  totalCommits: number
}

export async function fetchUserContributions(client: AxiosInstance, username: string) {
  const { data } = await client.get<ApiResponse<ContributionData>>(
    `/api/v1/repo/user-repos/${encodeURIComponent(username)}/contributions`
  )
  return data
}

export async function fetchExploreRepositories(
  client: AxiosInstance, 
  options: { search?: string; language?: string } = {}
) {
  const { search, language } = options
  let url = '/api/v1/repo/explore'
  const params = new URLSearchParams()
  if (search) params.append('search', search)
  if (language) params.append('language', language)
  if (params.toString()) url += `?${params.toString()}`

  const { data } = await client.get<ApiResponse<Repository[]>>(url)
  return data
}

export async function fetchRepository(client: AxiosInstance, owner: string, name: string) {
  const { data } = await client.get<ApiResponse<Repository>>(`/api/v1/repo/${encodeURIComponent(owner)}/${encodeURIComponent(name)}`)
  return data
}

export async function fetchRepositoryTree(
  client: AxiosInstance, 
  owner: string, 
  name: string, 
  branch: string, 
  path = ''
) {
  const url = path 
    ? `/api/v1/repo/${encodeURIComponent(owner)}/${encodeURIComponent(name)}/tree/${encodeURIComponent(branch)}/${path}`
    : `/api/v1/repo/${encodeURIComponent(owner)}/${encodeURIComponent(name)}/tree/${encodeURIComponent(branch)}`
  const { data } = await client.get<ApiResponse<TreeEntry[]>>(url)
  return data
}

export async function fetchBranches(client: AxiosInstance, owner: string, name: string) {
  const { data } = await client.get<ApiResponse<BranchInfo[]>>(
    `/api/v1/repo/${encodeURIComponent(owner)}/${encodeURIComponent(name)}/branches`
  )
  return data
}

export async function fetchBlobContent(
  client: AxiosInstance, 
  owner: string, 
  name: string, 
  branch: string, 
  path: string
) {
  const { data } = await client.get<ApiResponse<BlobResponse>>(
    `/api/v1/repo/${encodeURIComponent(owner)}/${encodeURIComponent(name)}/blob/${encodeURIComponent(branch)}/${path}`
  )
  return data
}

export async function fetchCommits(
  client: AxiosInstance, 
  owner: string, 
  name: string, 
  branch: string,
  page = 1,
  limit = 20
) {
  const { data } = await client.get<ApiResponse<Commit[]>>(
    `/api/v1/repo/${encodeURIComponent(owner)}/${encodeURIComponent(name)}/commits/${encodeURIComponent(branch)}?page=${page}&limit=${limit}`
  )
  return data
}

export async function fetchCommitDetail(
  client: AxiosInstance, 
  owner: string, 
  name: string, 
  hash: string
) {
  const { data } = await client.get<ApiResponse<Commit>>(
    `/api/v1/repo/${encodeURIComponent(owner)}/${encodeURIComponent(name)}/commit/${hash}`
  )
  return data
}

export async function fetchCommitDiff(
  client: AxiosInstance, 
  owner: string, 
  name: string, 
  hash: string
) {
  const { data } = await client.get<ApiResponse<DiffFile[]>>(
    `/api/v1/repo/${encodeURIComponent(owner)}/${encodeURIComponent(name)}/commit/${hash}/diff`
  )
  return data
}

export async function deleteRepository(client: AxiosInstance, owner: string, name: string) {
  const { data } = await client.delete<ApiResponse<void>>(`/api/v1/repo/${encodeURIComponent(owner)}/${encodeURIComponent(name)}`)
  return data
}

export async function updateRepository(
  client: AxiosInstance,
  owner: string,
  name: string,
  payload: {
    name?: string
    description?: string
    visibility?: 'public' | 'private'
    default_branch?: string
  }
) {
  const { data } = await client.patch<ApiResponse<Repository>>(
    `/api/v1/repo/${encodeURIComponent(owner)}/${encodeURIComponent(name)}`,
    payload
  )
  return data
}

export async function confirmDeleteRepository(
  client: AxiosInstance,
  owner: string,
  name: string,
  confirmationName: string
) {
  const { data } = await client.post<ApiResponse<void>>(
    `/api/v1/repo/${encodeURIComponent(owner)}/${encodeURIComponent(name)}/confirm-delete`,
    { confirmation_name: confirmationName }
  )
  return data
}

export async function toggleStar(client: AxiosInstance, owner: string, name: string) {
  const { data } = await client.post<ApiResponse<{ starred: boolean; stars: number }>>(
    `/api/v1/repo/${encodeURIComponent(owner)}/${encodeURIComponent(name)}/star`
  )
  return data
}

export async function checkStar(client: AxiosInstance, owner: string, name: string) {
  const { data } = await client.get<ApiResponse<{ starred: boolean; stars: number }>>(
    `/api/v1/repo/${encodeURIComponent(owner)}/${encodeURIComponent(name)}/star`
  )
  return data
}

export async function forkRepository(client: AxiosInstance, owner: string, name: string) {
  const { data } = await client.post<ApiResponse<Repository>>(
    `/api/v1/repo/${encodeURIComponent(owner)}/${encodeURIComponent(name)}/fork`
  )
  return data
}

export async function fetchStarredRepositories(client: AxiosInstance, username: string) {
  const { data } = await client.get<ApiResponse<Repository[]>>(`/api/v1/repo/user-repos/${encodeURIComponent(username)}/starred`)
  return data
}

// --- Pull Requests ---

export async function createPullRequest(
  client: AxiosInstance,
  owner: string,
  name: string,
  payload: { title: string; description: string; base_branch: string; compare_branch: string }
) {
  const formData = new FormData()
  formData.append('title', payload.title)
  formData.append('description', payload.description || '')
  formData.append('base_branch', payload.base_branch)
  formData.append('compare_branch', payload.compare_branch)

  const { data } = await client.post<ApiResponse<PullRequest>>(
    `/api/v1/repo/${encodeURIComponent(owner)}/${encodeURIComponent(name)}/pulls`,
    formData,
    { headers: { 'Content-Type': 'multipart/form-data' } }
  )
  return data
}

export async function fetchPullRequests(client: AxiosInstance, owner: string, name: string) {
  const { data } = await client.get<ApiResponse<PullRequest[]>>(`/api/v1/repo/${encodeURIComponent(owner)}/${encodeURIComponent(name)}/pulls`)
  return data
}

export async function fetchPullRequestDetail(client: AxiosInstance, owner: string, name: string, number: number) {
  const { data } = await client.get<ApiResponse<PullRequestDetail>>(`/api/v1/repo/${encodeURIComponent(owner)}/${encodeURIComponent(name)}/pulls/${number}`)
  return data
}

export async function mergePullRequest(client: AxiosInstance, owner: string, name: string, number: number) {
  const { data } = await client.post<ApiResponse<PullRequest>>(`/api/v1/repo/${encodeURIComponent(owner)}/${encodeURIComponent(name)}/pulls/${number}/merge`)
  return data
}
